### PR TITLE
HIVE-26768: HPLSQL UDF is not working if it is applied on a column of type varchar/char/decimal in a table.

### PR DESCRIPTION
**Pull request by mdayakar**
HPLSQL UDF is not working if it is applied on a column of type varchar/char/decimal in a table.

Checked and supported all primitive types like boolean, smallint, int, bigint, float, double, string, date, timestamp, decimal, varchar and char types.

### What changes were proposed in this pull request?
The changes supports users to create HPLSQL functions with all primitive types.

### Why are the changes needed?
To fix HIVE-26768 and also it helps users to create HPLSQL functions with all primitive types.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Tests are added.

### CODE REVIEW

**Comment by zeroflag** [comment]
LGTM

File: hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java [hunk 1]
```diff
@@ -128,27 +127,96 @@ public Object evaluate(DeferredObject[] arguments) throws HiveException {
...
+    case DOUBLE:
+      Double doubleValue = (Double) ((DoubleObjectInspector) objectInspector).getPrimitiveJavaObject(inputObject);
+      if (doubleValue != null) {
+        exec.setVariable(name, new Var(doubleValue));
+      }
+      break;
+    case STRING:
+      String strValue = ((StringObjectInspector) objectInspector).getPrimitiveJavaObject(inputObject);
+      if (strValue != null) {
+        exec.setVariable(name, strValue);
```
Comment (kasakrisz) [inline hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java:183]: Should the valued be wrapped in a `Var` instance?

File: hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java [hunk 2]
```diff
@@ -128,27 +127,96 @@ public Object evaluate(DeferredObject[] arguments) throws HiveException {
...
+    case DECIMAL:
+      HiveDecimal decimalValue = ((HiveDecimalObjectInspector) objectInspector).getPrimitiveJavaObject(inputObject);
+      if (decimalValue != null) {
+        exec.setVariable(name, new Var(decimalValue.bigDecimalValue()));
+      }
+      break;
+    case VARCHAR:
+      HiveVarchar varcharValue = ((HiveVarcharObjectInspector) objectInspector).getPrimitiveJavaObject(inputObject);
+      if (varcharValue != null) {
+        exec.setVariable(name, varcharValue.getValue());
```
Comment (kasakrisz) [inline hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java:209]: Should the valued be wrapped in a Var instance?

File: hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java [hunk 3]
```diff
@@ -128,27 +127,96 @@ public Object evaluate(DeferredObject[] arguments) throws HiveException {
...
+      }
+      break;
+    case CHAR:
+      HiveChar charValue = ((HiveCharObjectInspector) objectInspector).getPrimitiveJavaObject(inputObject);
+      if (charValue != null) {
+        exec.setVariable(name, charValue.getStrippedValue());
```
Comment (kasakrisz) [inline hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java:215]: Should the valued be wrapped in a Var instance?

File: itests/hive-unit/src/test/java/org/apache/hive/beeline/TestHplSqlViaBeeLine.java [hunk 4]
```diff
@@ -171,17 +171,197 @@ public void testPackage() throws Throwable {
...
+            "INSERT INTO result VALUES('Daya');\n" +
+            "INSERT INTO result VALUES('Alice');\n" +
+            "CREATE FUNCTION hello(c char(10))\n" +
+            "   RETURNS char(10)\n" +
+            "BEGIN\n" +
+            "   RETURN 'Hello ' || c || '!';\n" +
+            "END;\n" +
+            "SELECT hello(col_c) FROM result;\n";
+    testScriptFile(SCRIPT_TEXT, args(), "Hello Daya!.*Hello Alice!");
   }
```
Comment (kasakrisz) [inline itests/hive-unit/src/test/java/org/apache/hive/beeline/TestHplSqlViaBeeLine.java:365]: What (should) happen when the udf formal parameter type and the actual parameter parameter type is different?
Could you please add tests to cover that?
Ex.:
```
CREATE TABLE result (col_c int);
...
CREATE FUNCTION hello(i smallint)
...
``` 

```
CREATE TABLE result (col_c date);
...
CREATE FUNCTION hello(i double)
...
```

File: hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java (hunk 2 above)
Comment (mdayakar) [reply hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java:209]: Internally Exec code Var object is getting created for string type input.
Refer https://github.com/apache/hive/blob/master/hplsql/src/main/java/org/apache/hive/hplsql/Exec.java#L210

File: hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java (hunk 3 above)
Comment (mdayakar) [reply hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java:215]: Internally Exec code, Var object is getting created for string type input.
Refer https://github.com/apache/hive/blob/master/hplsql/src/main/java/org/apache/hive/hplsql/Exec.java#L210

File: itests/hive-unit/src/test/java/org/apache/hive/beeline/TestHplSqlViaBeeLine.java (hunk 4 above)
Comment (mdayakar) [reply itests/hive-unit/src/test/java/org/apache/hive/beeline/TestHplSqlViaBeeLine.java:365]: The function will be called if the actual parameter type can be converted to udf parameter type. For example, if UDF is taking String and actual parameter is decimal then in this case UDF will be called as Decimal can be easily converted to String. Anyway I will add a testcase for the same.

File: hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java (hunk 1 above)
Comment (mdayakar) [reply hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java:183]: Internally Exec code, Var object is getting created for string type input.
Refer https://github.com/apache/hive/blob/master/hplsql/src/main/java/org/apache/hive/hplsql/Exec.java#L210

File: hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java (hunk 2 above)
Comment (mdayakar) [reply hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java]: Fixed

File: hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java (hunk 3 above)
Comment (mdayakar) [reply hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java]: Fixed

File: itests/hive-unit/src/test/java/org/apache/hive/beeline/TestHplSqlViaBeeLine.java (hunk 4 above)
Comment (mdayakar) [reply itests/hive-unit/src/test/java/org/apache/hive/beeline/TestHplSqlViaBeeLine.java:365]: Fixed

File: hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java (hunk 1 above)
Comment (mdayakar) [reply hplsql/src/main/java/org/apache/hive/hplsql/udf/Udf.java]: Fixed

**Review by kasakrisz — APPROVED** [review]

**Pull request merged**